crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3983,6 +4017,116 @@ impl<'db> TypeInferenceBuilder<'db> {
             }
         }
     }
+
+    fn infer_subscript_type_expression(
+        &mut self,
+        subscript: &ruff_python_ast::ExprSubscript,
+        value_ty: Type<'db>,
+    ) -> Type<'db> {
+        let ast::ExprSubscript {
+            range: _,
+            value: _,
+            slice,
+            ctx: _,
+        } = subscript;
+
+        match value_ty {
+            Type::Instance(InstanceType {
+                class: _,
+                known: Some(known_instance),
+            }) => self.infer_parameterized_known_instance_type_expression(known_instance, slice),
+            _ => {
+                self.infer_type_expression(slice);
+                Type::Todo
+            }
+        }
+    }
+
+    fn infer_parameterized_known_instance_type_expression(
+        &mut self,
+        known_instance: KnownInstance,
+        parameters: &ast::Expr,
+    ) -> Type<'db> {
+        match known_instance {
+            KnownInstance::Literal => {
+                let Some(slice_ty) = self.infer_literal_parameter_type(parameters) else {
+                    self.diagnostics.add(
+                        parameters.into(),
+                        "invalid-literal-parameter",
+                        format_args!(""),
```

Comment:
I think we do need *some* kind of message here to the user telling them what went wrong :-)

Reply:
Yeah I misunderstood this comment and removed the message.
https://github.com/astral-sh/ruff/pull/13874#discussion_r1819898105
Was the previous message good?

Reply:
yes, the previous message was fine! We should _have_ a message here, we just shouldn't _assert_ the message in our tests (because it's unnecessary).

For example, take our `cyclic-class-def` error code: we emit a message here, and it will be shown to users:

https://github.com/astral-sh/ruff/blob/b7e32b0a18487e8c9813e0d189d7a5d625ff5c3f/crates/red_knot_python_semantic/src/types/infer.rs#L481-L489

but we don't assert the full message in our tests, because it's the same every time, and it would just be a pain to update all those tests if we ever changed the error message slightly. Instead, we just assert the error code:

https://github.com/astral-sh/ruff/blob/b7e32b0a18487e8c9813e0d189d7a5d625ff5c3f/crates/red_knot_python_semantic/resources/mdtest/mro.md?plain=1#L350-L351